doc/DesignPrinciples.md

```diff
@@ -0,0 +1,236 @@
+* `isactive(rws, i::Integer)` - at various stages of `knuthbendix` procedure to check if the rule is active;
+* `setinactive!(rws, i::Integer)` - during `deriverule!`, to mark rules as inactive.
+
+### `KnuthBendix.RewritingSystem`
+
+We implemented a simple `RewritingSystem{W<:AbstractWord, O<:WordOrdering} <:AbstractRewritingSystem{W}`
+which stores the defining ordering in one of its fields.
+Each rule (in the current implementation) consist of an ordered pair of words `(lhs, rhs)::Tuple{W,W}`
+(i.e. `Base.lt(ordering(rws), lhs, rhs)` holds), which represents rewriting rule
+> `lhs` → `rhs`.
+
+**IDEA:** create a distinct type/class for a single rewriting rule and store the rules as a list of this objects.
+Such rule could store also its status (active or not), etc. See #24.
+
+Inside `RewritingSystem{W}` we also store `rules::Vector{W}` and (paired with it) `act::BitArray`,
+which stores `1` when the corresponding rule in `rules` is active and `0` otherwise.
+Note that iterating over `KnuthBendix.rules(rws)` returns all rules, regardless of their active status.
+
+**TODO:** return an specialized iterator over rules of `rws` which iterates only over active rules.
+This would surely make code simpler/easier to read and more generic.
+
+### Simplifying rules
+
+Method `simplifyrule!(lhs::AbstractWord, rhs::AbstractWord, A::Alphabet)` was suggested in the book by Sims.
+It takes a rule (typically prodcued during Knuth-Bendix procedure) and (as of now)
+cancels out maximal invertible prefix and suffix of both sides of the rule.
+
+## States and Automata
+
+### States
+
+Before describing automata itself we should describe a single state in it.
+`KnuthBendix.State{N, W}<:KnuthBendix.AbstractState` struct is indexed by two parameters:
+* `N` the length of the alphabet we use (a letter and its inverse are considered as two distinct elements of the alphabet), and
+* `W` particular type of `AbstractWord` that we use.
+
+#### Details of implementation
+
+A state has a `name`: a word corresponding to the shortest path from initial state to the state in mention
+(that way it is easy to define a length function for a state - as mentioned in Sims - this is just a length of the name).
+
+A field `terminal` is a boolean which indicates whether the state represents a word,
+which is a left-hand side of some rewriting rule in `RewritingSystem`.
+In case this is the case (field `terminal` is `true`) field `rrule` should be equal to the `Word` representing
+the right-hand side of that rewriting rule.
+
+Field `ined` (incoming edges) is a vector of instances of `State` struct - those instances of `State`
+from which there is an edge leading to a state in mention
+(so for a `State` with name "abc" there must be - in particular - a `State` with name "ab" in the `ined` list).
+What is more: always on the first place on that list should be a state representing (having field `name`)
```

Comment:
why is this first place so important?

Reply:
It is not a requirement per se - this is just how things are with `makeindexautomaton!` (if you think about how this algorithm works, you will see that this condition is always assured). I think it is nevertheless a condition worth mentioning, as it may be useful if another take at automata will be tried (e.g. automata storing only simple paths).